README.md
<img src="https://raw.githubusercontent.com/jmscslgroup/strym/master/strym.png" alt="Strym Logo" align="center"/>


# Strym
[![Build Status](https://travis-ci.com/jmscslgroup/strym.svg?branch=master)](https://travis-ci.com/jmscslgroup/strym)
[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://GitHub.com/jmscslgroup/strym/graphs/commit-activity)
[![made-with-python](https://img.shields.io/badge/Made%20with-Python-1f425f.svg)](https://www.python.org/)
[![made-with-sphinx-doc](https://img.shields.io/badge/Made%20with-Sphinx-1f425f.svg)](https://www.sphinx-doc.org/)
[![PyPI version shields.io](https://img.shields.io/pypi/v/strym.svg)](https://pypi.python.org/pypi/strym/)
[![PyPI license](https://img.shields.io/pypi/l/strym.svg)](https://pypi.python.org/pypi/strym/)
[![Downloads](https://pepy.tech/badge/strym)](https://pepy.tech/project/strym)



__A python package for real-time CAN data logging, analysis and visualization to work with USB-CAN Interface.__

__Strym__ is a python package that provides APIs to interface with COMMA.AI panda for logging CAN data, analysis and visualization in real-time from supported modern vehicles such Toyota RAV4 and Honda Pilot. There are two kinds of functionality that __Strym__ provides: 

1. Real-time visualization of CAN data through comma.ai Panda and Giraffe connector.
2. Offline analysis and visualization of CAN Data from a CSV Formatted file.


## Philosophy behind Strym

Strym data is capable of handling timeseries data obtained from Comma.ai Panda and Giraffe Connector. 
Most functions and methods in `strym` expects timeseries data of following format

|   | Time               | Message |
|---|--------------------|---------|
| 1 | 1582056042.5040324 | 2.0     |
| 2 | 1582056043.5040324 | 2.1     |
| 3 | 1582056044.5040324 | 2.12    |
| 4 | 1582056045.5040324 | 1.98    |
| 5 | 1582056046.5040324 | 1.6     |

Here, data should be of type Pandas.DataFrame with two columns: Time and Message. 

However, scope of strym is not limited to timeseries data obtained from comma.ai Panda. Any timeseries data of above format is capable of harnessing methods available in `strym`.

## Quick Start for CAN Data Analysis and Visualization

You can use __Strym__ for quick visualization by importing `strymread`:
```python
import strym
from strym import strymread
from strym import ranalyze
import matplotlib.pyplot as plt
import pandas as pd
from pylab import rcParams
import strym.DBC_Read_Tools as dbc
import numpy as np
plt.rcParams["figure.figsize"] = (16,8)
rcParams.update({'font.size': 40})
dbcfile = '/home/ivory/VersionControl/Jmscslgroup/strym/examples/newToyotacode.dbc'
r =strymread(csvfile="/home/ivory/CyverseData/JmscslgroupData/PandaData/2020_02_18/2020-02-18-13-00-42-209119__CAN_Messages.csv", dbcfile=dbcfile)
```

Checkout documentation at https://jmscslgroup.github.io/strym/getting_started.html for more in-depth tutorials.

## Software Requirements
- Ubuntu 18.04 (not tested on any other version of Ubuntu, but might work)
- Python 3.x

### Note about installation on RASPBERRY PI for CAN Data Logging
If you are going to install the package on RASPBERRY PI, I highly recommend installing Python 3.7 from the source as there is no Py3.7 release for Raspberry PI.
You will also need to install pre-compiled binaries for NumPy otherwise you may encounter huge inconvenience while building NumPy wheels for Raspberry PI.


## Hardware Requirements for CAN Logging
- comma.ai CAN-USB __Panda__ board.
- comma.ai Giraffee Connector
- A modern vehicle with CAN Bus available such as Toyota RAV4, Toyota CHR, etc. 

## Installation Instructions

[![Install Instruction](https://img.youtube.com/vi/w2p1uYmHBPA/0.jpg)](https://www.youtube.com/watch?v=w2p1uYmHBPA&t=5s)

### Install Python

Install Python 3, either through anaconda or using the Ubuntu package manager. Alternatively, you can also build Python 3.7 from source as explained below:

```bash
sudo apt-get update -y
sudo apt-get install build-essential tk-dev libncurses5-dev libncursesw5-dev libreadline6-dev libdb5.3-dev libgdbm-dev libsqlite3-dev libssl-dev libbz2-dev libexpat1-dev liblzma-dev zlib1g-dev libffi-dev -y

wget https://www.python.org/ftp/python/3.7.2/Python-3.7.2.tar.xz
tar xf Python-3.7.2.tar.xz
cd Python-3.7.2
./configure
make -j 4
sudo make altinstall
```

I recommend using python's virtual environment for python package installation. For the sake of following instructions, let's assume that you are using the `virtualenv`  package to create a python virtual environment. 

```bash
sudo apt install virtualenv
```
First, create a directory where your virtual environment folder will reside.

```bash
mkdir ~/VirtualEnv
```
Now, we will create a python virtual environment using python3.7. Let's name the virtual environment *stream*.

```bash
cd ~/VirtualEnv
virtualenv --python=python3.7 stream
```

Activate the virtual environment by typing:

```bash
source ~/VirtualEnv/stream/bin/activate
```

### Install strym

`pip install strym`

This will install the strym package in your `stream` virtual environment.

Now you are ready to use __Strym__.

## Usage for Real-Time Visualization of CAN messages using Strym

Plug your Comma AI Panda device using Giraffe Connector to your CAR's OBD port for data logging and streaming. Insert one end of the USB to Panda Device and another end to your laptop.

In python, you will be required to create an object of type `Strym`:


See `strym_impl.py` for one such usage example in the [example folder](https://github.com/jmscslgroup/strym/blob/master/examples), however, I am provided details of an example below:


Create a new file. I will use the gedit to create a new file. You will be required to pass a path of the CAN Database DBC file to `strym` while instantiating its object. Once you have a `strym` object, you can call its `isoviz()` function. `isoviz()` function takes two arguments: i) the message type that you want to visualize, e.g. SPEED ii) attribute number to plot specific signal of the desired message type. `isoviz()` function will simultaneously capture CAN message in a CSV file and also plot the desired message's signal. To terminate, press CTRL-C. Upon pressing CTRL-C, a SIGINT signal handler will be called that will terminate the logging of CAN messages and also save a matplotlib figure of the desired message's signal in pdf and pickle format.

```bash
gedit viz_example.py
```

```python
from strym import strym
import cantools
import sys, math, time
import signal

db = './newToyotacode.dbc'

Viz = strym(dbcfile = db)

message_type_to_visualize = 'SPEED'
message_attribute_number_to_visualize = 1

visualize = False
options = {"log": "info" }
Viz.isolog(visualize, message_type_to_visualize, message_attribute_number_to_visualize,  **options)

signal.signal(signal.SIGINT, Viz.kill)

print('Datafile saved is {}'.format(Viz.logfile))


```

You will need a DBC file to parse can messages. Download an example DBC file [here](https://github.com/jmscslgroup/strym/blob/master/examples/newToyotacode.dbc)

To run the above program:

```bash
source ~/VirtualEnv/stream/bin/activate
python viz_example.py`
```

If you are done, press CTRL-C.

## Issues
If you run into any issues, please use the issue feature of GitHub to log your issues. I will try my best to address any issue as soon as
possible.

For an issue related to installation/use on windows, please see following filed issues:

1. [Issue #8: OSError: [WinError 126]](https://github.com/jmscslgroup/strym/issues/8)

## Contributing to this project
If you like to contribute to this project, please fork this repository to your GitHub account, create a new branch for yourself and send a pull request for the merge. After reviewing the changes, we will decide if this is a good place to add your changes.

## Authors and Contributors
- Rahul Bhadani ( rahulbhadani@email.arizona.edu)
- Jonathan Sprinkle (sprinkjm@email.arizona.edu)
- Gustavo Lee (gustavolee@email.arizona.edu)
- Matthew Nice (matthew.nice@vanderbilt.edu)
- George Gunter (gunter.gl@gmail.com)
- Safwan Elmadani (safwanelmadani@email.arizona.edu)

## Licensing

    License: MIT License 
    Copyright 2019-2020 Rahul Bhadani, Jonathan Sprinkle, Arizona Board of Regents
    Initial Date: Nov 12, 2019
    Permission is hereby granted, free of charge, to any person obtaining 
    a copy of this software and associated documentation files 
    (the "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to 
    permit persons to whom the Software is furnished to do so, subject 
    to the following conditions:

    The above copyright notice and this permission notice shall be 
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF 
    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
    PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT 
    SHALL THE AUTHORS, COPYRIGHT HOLDERS OR ARIZONA BOARD OF REGENTS
    BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN 
    AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE 
    OR OTHER DEALINGS IN THE SOFTWARE.




CHANGELOG.rst
=========
Changelog
=========

0.4.0 - 2021-Feb-02
---------------------
- ``platoons`` module for analysis using multiple driving routes
- Currently supporting finding similar routes using ``spatial_finder`` functions

0.3.6 - 2020-Nov-11
--------------------
- Interactive plot in ``strymread``
- Default map for ``strymmap`` set to Mapbox with an option for Googlemap
- Now ``strymread`` function ``plt_ts`` can take a non-default column other than 'Message' for plotting as an argument
- Bug fixes


0.3 - 2020-Oct-15
--------------------
- Bug fixes in ``strymread``
- Strym tools
    - ``graham_scan`` function to find convex hull and its perimeter for a cluster
    - ``ellipse_fit`` function to fit an ellipse/circle to a cluster of points on phase-space

0.2.4 - 2020-Sep-26
--------------------
- Documentation bug fixes
- Plotting bug fixes and improvement
- Strym tools
    - ACD Tool added to measure stop-and-go distance. Animation generation for phasespace evolution supported
    - Acceleration messages added to :code:`strymread` to be dbc independent
    - DBC now supports headlight message for highbeam, lowbeam etc. 

0.2.2 - 2020-Sep-09
--------------------
- Documentation format fix
- Minor fixes in code

0.2.1 - 2020-Aug-17
--------------------
- Documentation bug fixes
- Minor improvment over state-space calculation

0.2.0 - 2020-Aug-15
-----------------------
- Standalone functions from all classes moved to respective class definition as static method to reflect their usability
- Sphinx version updated
- Strym now packages DBC file. Currently suppported DBC file. Currently supported DBC file: Honda Pilot 2017, Toyota RAV4 2019 and 2020
- Support for creating serverless sqlite3 database in local system for easy querying of driving information

- class: :code:`strymread`
    - Bug fix: removes duplicate timestamped data from signals
    - `get_ts` now has additional column *BUS* that provides number of each timestamped message
    - Option to save RAW, UNDECODED Dataframe as sqlite3 via `createdb` parameter while object instantiation
    - A helper message `lead_distance` to retrive lead distance from csv file
    - Function name `extract` changed to `export2mat` that creates matlab .mat file of known, import signals
    - `state_space` function to extract consolidated dataframe with a number of signals that represents state space of the vehicle, resampled to desire frequency. Optional argument todb, when `True` saves state space dataframe to a sqlite3 table `STATE_SPACE`
    - `differentiate` method automatically falls back to AutoEncoder method when number of datapoints is less than 6.

0.1.13 - 2020-July-29
----------------------
- class: :code:`strymread`
    - Fixed Issue #16

0.1.12 - 2020-July-25
----------------------
- class: :code:`strymread`
    - Fixed a scaling bug in prediction made in AE-based differentiation

0.1.11 - 2020-July-23
----------------------
- Changes to permit strymread to work with Honda
 
- class :code:`strymread`
    - Added rel_velocity to retrieve relative velocity of targets from radar traces
    - Added create_chunks function to split discontinuous timeseries into continuous-chunked timeseries
    - Autoencoder based denoising method and application for estimation of lead vehicle's velocity
    - DBC support for Honda Pilot
    - Some inconsistency fixes

2020-July-02
---------------------
- modified for robust calling of meta and dashboard from snakemake files

- class :code:`meta`
    - Takes an array of dbc files, and calls methods from the vin_parser package to confirm that the VIN of a CSV file shoudl correspond to a particular dbc

- class :code:`strymread`
    - Added new functionality to do dictionary lookup of message/signal pairs when extracting a timeseries, rather than assuming all message/signal pairs correspond to Toyota naming conventions. New methods to touch when adding new message/signal pairs are prefixed with `_dbc_`

0.1.10 - 2020-July-01
---------------------
- setup modified to package examples and README folder in install path

- class :code:`dashboard`
    - works within the strym package to collect metadata files from within a folder and print interesting aspects of the collection
    
- class :code:`meta`
    - Works within the strym package to extract metadata from drives that are recorded using libpanda, with optional corresponding dashcam video

- class :code:`strymread`
    - Checks for monotonicity of the time in the recorded data
    - Two new Flag Attributes added: burst and success. Burst flag tells if data was recorded in burst, success tells if reading of csv file was successful

- class :code:`strymmap`
    - Conditionally import of bokeh and other widget-based libraries in Jupyter only to making strym work from terminal
    
    
0.1.9 - 2020-May-20
-------------------
- class :code:`strymread`
    - A new function :code:`extract` for extracting data in .MAT format to work with matlab
    - Changes to :code:`msg_subset` function, now returns an object of type :code:`strymread` with modified dataframe
    - Changes to plotting functions

0.1.8 - 2020-May-16
-------------------

- A few bug fixes and cosmetic changes.
- Freezing the master branch and creating a devel branch for active development.


strym/strym.py
#!/usr/bin/env python
# coding: utf-8

# Author : Rahul Bhadani
# Initial Date: Nov 11, 2019
# About: strym class uses comma.ai panda package to capture can data from comma.ai panda device
#   and plot in the real time. Read associated README for full description
# License: MIT License


#   Permission is hereby granted, free of charge, to any person obtaining
#   a copy of this software and associated documentation files
#   (the "Software"), to deal in the Software without restriction, including
#   without limitation the rights to use, copy, modify, merge, publish,
#   distribute, sublicense, and/or sell copies of the Software, and to
#   permit persons to whom the Software is furnished to do so, subject
#   to the following conditions:

#   The above copyright notice and this permission notice shall be
#   included in all copies or substantial portions of the Software.

#   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
#   ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
#   TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
#   PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
#   SHALL THE AUTHORS, COPYRIGHT HOLDERS OR ARIZONA BOARD OF REGENTS
#   BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
#   AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
#   OR OTHER DEALINGS IN THE SOFTWARE.



__author__ = 'Rahul Bhadani'
__email__  = 'rahulbhadani@email.arizona.edu'
__version__ = "0.0.0" # this is set to actual version later
# For System and OS level task
import sys, getopt

## General Data processing and visualization Import

import struct
import signal
import binascii
import bitstring
import time
import datetime
import serial
import csv
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from matplotlib import cm
import pandas as pd # Note that this is not commai Panda, but Database Pandas
import matplotlib.animation as animation
from matplotlib import style
from matplotlib.ticker import LinearLocator, FormatStrFormatter
import uuid
import scipy.special as sp
import pickle
import os
from os.path import expanduser
from packaging import version
import warnings

try:
    import libusb1
except OSError:
    warnings.warn("libusb-1.0.dll may not be present. Normal strymread operations to read CAN data from csv files won't be affect, but you won't be able to read live data from car. For resolution to this issue, follow the steps described in https://github.com/jmscslgroup/strym/issues/8#issuecomment-652539765")

try:
    import usb1
except OSError:
    warnings.warn("libusb-1.0.dll may not be present. Normal strymread operations to read CAN data from csv files won't be affect, but you won't be able to read live data from car. For resolution to this issue, follow the steps described in https://github.com/jmscslgroup/strym/issues/8#issuecomment-652539765")

# cantools import
import cantools

from pathlib import Path
version_src = ''
try:
    import importlib.resources as pkg_resources
    with pkg_resources.path('strym', 'version') as rsrc:
        version_src = rsrc
except ImportError:
    # Try backported to PY<37 `importlib_resources`.
    print("Python older than 3.7 detected. ")
    try:
        import importlib_resources as pkg_resources
        with pkg_resources.path('strym', 'version') as rsrc:
            version_src = rsrc
    except ImportError:
        print("importlib_resources not found. Install backported importlib_resources through `pip install importlib-resources`")

v = Path(version_src).open(encoding = "utf-8").read().splitlines()
__version__ = v[0].strip()

def timeout(func, args=(), timeout_duration=2, default=None, **kwargs):
    """This spwans a thread and runs the given function using the args, kwargs and
    return the given default value if the timeout_duration is exceeded
    """
    import threading

    class InterruptableThread(threading.Thread):
        def __init__(self):
            threading.Thread.__init__(self)
            self.result = default

        def run(self):
            try:
                self.result = func(*args, **kwargs)
            except:
                pass

    it = InterruptableThread()
    it.start()
    it.join(timeout_duration)
    return it.result

def get_latest_strym_version():
    from subprocess import check_output, CalledProcessError

    try:  # needs to work offline as well
        result = check_output(["yolk", "-V", "strym"])
        return result.split()[1].decode("utf-8")
    except CalledProcessError:
        return "0.0.0"


def check_for_latest_version():

    latest_version = timeout(
        get_latest_strym_version, timeout_duration=5, default="0.0.0"
    )
    if version.parse(__version__) < version.parse(latest_version):
        
        warnings.warn("{}\n{}\n{}\n{}\n{}\n{}".format(
            "There is a newer version of strym available on PyPI:\n",
            "Your version: \t",
            __version__,
            "Latest version: \t",
            latest_version,
            "Consider updating it by using command pip install --upgrade strym"
        ))


check_for_latest_version()

class strym:
    '''
    `strym`  class records data from Comm AI Panda and visualize in real time.
    The constructor first gets an "USB context" by creating  `USBContext` instance.
    Then, it browses available USB devices and open the one whose manufacturer is
    COMMA.AI. One right device is identified,  `strym` creates a device handle,
    enables automatic kernel driver detachment and claim interface for I/O operation.

    Read and Write for USB devices are either done synchronosly or in isochronos mode.

    If your interest is merely in capturing data, you should perform synchronous mode.

    For (almost) real time visualization, isochronous mode is the way to go.

    Parameters
    -------------
    dbcfile: `string`
    Provide path of can database file in order to decode the message

    kwargs:
        Arbitrary keyword arguments.

        path: `string`
            Specify the path/folder where data will be saved. By default path is set to  `~/CyverseData/JmscslgroupData/PandaData`

    See Also
    -----------------

    ##  https://github.com/gotmc/libusb

    ## https://pypi.org/project/libusb1/

    ## https://vovkos.github.io/doxyrest/samples/libusb/index.html

    ## https://github.com/vpelletier/python-libusb1

    ## https://www.beyondlogic.org/usbnutshell/usb4.shtml

    '''
    def __init__(self, dbcfile, **kwargs):

        # Get the home folder of the current user
        home = expanduser("~")

        # Create a folder CyverseData where all the log files will be record.
        self.data_folder = home+ '/CyverseData/JmscslgroupData/PandaData'

        ## Parse the variable number of arguments

        try:
            self.data_folder = kwargs["path"]
        except KeyError as e:
            pass

        # Get the USB Context
        self.context = usb1.USBContext()
        # Get all the USB device list
        device_list =  self.context.getDeviceList()
        commaai_device = None

        # Iterate over the list of devices
        for device in device_list:
            try:
                device_manufacturer = device.getManufacturer()
                print('Device manufacturer is {}\n'.format(device_manufacturer))
                if device_manufacturer == 'comma.ai':
                    commaai_device = device
                    print("We found a COMMA AI Device with serial number {}".format(commaai_device.getSerialNumber()))
                    break
            except usb1.USBErrorAccess:
                # If the device is not accessible, do not do anything
                # print('USB Device Not accessible')
                pass

        if commaai_device is None:
            print("No comma.ai device was found. Aborting")
            sys.exit(-1)

        self.device = commaai_device
        # Save the serial number for future use
        self.serial = commaai_device.getSerialNumber()

        # open the comma.ai device and obtain a device handle. A handle allows you to
        # perform I/O on the device in question. Internally, this function adds a
        # reference to the device and makes it available to you through
        # `libusb_get_device()`. This reference is removed during libusb_close().
        # This is a non-blocking function; no requests are sent over the bus.
        self.handle = commaai_device.open()

        # set_auto_detach_kernel_driver to  enable/disable libusb's automatic kernel driver detachment.
        self.handle.setAutoDetachKernelDriver(True)

        # You must claim the interface you wish to use before you can perform I/O on any of its endpoints.
        self.handle.claimInterface(0)

        # define endpoint for reading
        self.ENDPOINT_READ = 1
        # buffer size
        self.BUFFER_SIZE = 16
        # dbc file from constructor
        self.dbcfile = dbcfile
        # load can database from dbc file
        self.db = cantools.database.load_file(dbcfile)

        # Set up the figure
        self.fig = plt.figure()
        self.axis = self.fig.add_subplot(1,1,1)

        # logfile name attribute, initially None, it will be given value when we are ready to log the message
        self.csvwriter = None
        # Variable to Hold Specified Data for visualization
        self.data = []
        # Variable to Hold Time
        self.time = []
        # Boolean flag to keep recording data
        self.keep_recording = True

        # Message Type and attributes will be saved into these variables. This is only useful when you want to visualize the specific data
        self.msg_type = None
        self.attribute_num = None
        self.attribute_name = None
        self.newbuffer = None

    def process_received_data(self, transfer):
        '''
        `process_received_data` function implements a callback that processes the reeceived data
        from USB in isochronous mode.  Once data is extracted from buffer, it is saved in the object's data variable.
        The data is used to update the plot in the real time.

        '''
        curr_time = time.time() # Records time of collection

        if transfer.getStatus() != usb1.TRANSFER_COMPLETED:
            # Transfer did not complete successfully, there is no data to read.
            # This example does not resubmit transfers on errors. You may want
            # to resubmit in some cases (timeout, ...).
            return
        self.newbuffer = transfer.getBuffer()[:transfer.getActualLength()]

        if self.newbuffer is None:
            return

        # parse the can buffer into message ID, message, and bus number
        can_recv = self.parse_can_buffer(self.newbuffer)

        this_message = None
        this_message_name = None
        for message_id, _, new_message, bus  in can_recv:
            if self.typelist is not None and message_id not in self.typelist:
                continue

            self.csvwriter.writerow(([str(curr_time), str(binascii.hexlify(self.newbuffer).decode('utf-8'))  , str(bus), str((message_id)), str(binascii.hexlify(new_message).decode('utf-8')), len(new_message)]))
            if self.visualize:
                try:
                    this_message = self.db.get_message_by_frame_id(message_id)
                    this_message_name = this_message.name

                    # if the message currently received is in the list of messageTypes to be plotted, parse it and plot it
                    match_bool = False

                    if self.match == "exact":
                        match_bool = self.msg_type == this_message_name
                    elif self.match == "in":
                        match_bool =  self.msg_type in this_message_name

                    if match_bool :
                        decoded_msg = self.db.decode_message(this_message_name, bytes(new_message))
                        attribute_names = list(decoded_msg.keys())
                        self.attribute_name = attribute_names[self.attribute_num]
                        data =decoded_msg[self.attribute_name]
                        print('Time: {}, Data: {}'.format(curr_time, data))
                        self.data.append(data)
                        self.time.append(curr_time)

                        # Only plot 500 points at a time
                        # Check if data doesn't have 500 points then consume all of the data
                        if len(self.data) > 500:
                            data500 = self.data[-500:]
                            time500 = self.time[-500:]
                        else:
                            data500 = self.data
                            time500 = self.time

                        self.axis.clear()
                        self.axis.plot(time500, data500, linestyle='None', color='firebrick', linewidth=2, marker='.', markersize = 3)
                        self.axis.set_axisbelow(True)
                        self.axis.minorticks_on()
                        self.axis.grid(which='major', linestyle='-', linewidth='0.5', color='salmon')
                        self.axis.grid(which='minor', linestyle=':', linewidth='0.25', color='dimgray')
                        plt.title(self.msg_type + ": " +  self.attribute_name)
                        plt.xlabel('Time')
                        plt.ylabel(self.attribute_name)
                        self.axis.plot()
                        plt.draw()
                        plt.pause(0.00000001)
                except KeyError as e:
                    # print("this_message_name: {}".format(this_message_name))
                    if  self.log == "debug":
                        print('Message ID not supported by current DBC files ["{}"]' .format(e))
                    continue


    def _visualize(self ):
        '''
        This is internal function meant to visualize specific attribute of the given message passed to
        `isolog` function.

        '''
        pass

    def isolog(self, visualize, msg_type, attribute_num, **kwargs):
        '''
        `isoviz()` function will log everything in asynchronous manner but only visualize specific attribute of the given message.
        Upon pressing ctrl-C, the logging will terminate and SIGINT signal handler
        will create a plot and save in two formats: python's pickle format and pdf.

        `isoviz` is responsible handling data transfer in the isochronous mode and parsing through callback function `process_received_data`

        See https://vovkos.github.io/doxyrest/samples/libusb/group_libusb_asyncio.html?highlight=transfer#details-group-libusb-asyncio
        for more detail

        Parameters
        -------------

        visualize: `bool`
            specifies whether to visaulize while logging the CAN data

        msg_type: `string`
            specifies a valid message type from the DBC file

        attribute_num: `int`
            select the specific attribute from the given `mgs_type` to be displayed

        **kwargs:
            Arbitrary keyword arguments.

            log: `enumeration: {info, debug}`
                set log level to info and debug

            match: `enumeration: {exact, in}`
                how the message type and specified attribute should be matched for visualization. `exact` specifies exact match, `in` specifies substring matching.

        '''
        self.msg_type = msg_type
        self.attribute_num = attribute_num
        self.visualize = visualize

        self.log = "info"
        try:
            self.log = kwargs["log"]
        except KeyError as e:
            #print("KeyError: {}".format(str(e)))
            #raise
            pass

        self.match = "exact"
        try:
            self.match = kwargs["match"]
        except KeyError as e:
            #print("KeyError: {}".format(str(e)))
            #raise
            pass

        self.typelist  = None
        try:
            self.typelist = kwargs["typelist"]
        except KeyError as e:
            #print("KeyError: {}".format(str(e)))
            #raise
            pass

        dt_object = datetime.datetime.fromtimestamp(time.time())

        # Now create a folder inside CyverseData corresponding to today's date.
        todaysfolder = dt_object.strftime('%Y_%m_%d')

        path = self.data_folder  + "/" + todaysfolder
        if not os.path.exists(path):
            os.makedirs(path)

        dt = dt_object.strftime('%Y-%m-%d-%H-%M-%S-%f')
        logfile = path + '/' + dt + '_'   + '_CAN_Messages'+'.csv'
        self.logfile = logfile
        filehandler = open(logfile, 'a')
        print('Writing data to file: '+logfile)
        print('Press Ctrl - C to terminate')
        self.csvwriter = csv.writer(filehandler)
        self.csvwriter.writerow(['Time','Buffer','Bus', 'MessageID', 'Message', 'MessageLength'])

        while self.keep_recording:
            try:
                # Get an `USBTransfer` instance for asynchronous use.
                transfer = self.handle.getTransfer()
                transfer.setBulk(usb1.ENDPOINT_IN | self.ENDPOINT_READ, self.BUFFER_SIZE, callback = self.process_received_data,)
                try:
                    transfer.submit()
                except usb1.DoomedTransferError:
                    pass

                try:
                    self.context.handleEvents()
                except usb1.USBErrorInterrupted:
                    pass

            except KeyboardInterrupt as e:
                # Capture the SIGINT event and call plot function to finalize the plot and save the data
                self.kill(signal.SIGINT)

    #signal.signal(signal.SIGINT, self.kill)


    # SIGINT signal handler that will terminate lself.axogging of can data and save a final plot of the desired attribute of a message type

    def kill(self, sig):
        """

        `kill` catches SIGINT or CTRL-C while recording the data
        and closes the comma ai device connection

        """
        self.handle.close()
        print('CTRL-C (SIGINT) received. Stopping log.')
        self.keep_recording = False

        if self.msg_type is None:
            self.msg_type = 'Message Type'

        if self.attribute_num is None:
            self.attribute_num = 'Attribute'

        if self.visualize:
            # Ctrl-C Also saves the current figure being visualized with all data plotted on it.
            self.axis.clear()
            plt.rcParams["figure.figsize"] = (16,8)
            self.axis.plot(self.time, self.data, linestyle='None', color='firebrick', linewidth=2, marker='.', markersize = 3)
            self.axis.set_axisbelow(True)
            self.axis.minorticks_on()
            self.axis.grid(which='major', linestyle='-', linewidth='0.5', color='salmon')
            self.axis.grid(which='minor', linestyle=':', linewidth='0.25', color='dimgray')
            plt.title(self.msg_type + ": " + self.attribute_name)
            plt.xlabel('Time')
            plt.ylabel(self.attribute_name)
            current_fig = plt.gcf()
            file_name_to_save = self.logfile[0:-4]
            current_fig.savefig(file_name_to_save + ".pdf", dpi = 300)
            pickle.dump(self.fig,open(file_name_to_save + ".pickle",'wb'))


    def parse_can_buffer(self, dat):
        """
        `parse_can_buffer` parses the can data received through the USB device
        and returns list of message ID, message and bus number

        Parameters
        -------------
        dat: `bytearray`
            byte data to be  parsed

        Returns
        ------------
        `list`
            Returns a list containing message ID, message and bus number
        """
        ret = []
        for j in range(0, len(dat), 0x10):
            ddat = dat[j:j+0x10]
            f1, f2 = struct.unpack("II", ddat[0:8])
            extended = 4
            if f1 & extended:
                address = f1 >> 3
            else:
                address = f1 >> 21
            dddat = ddat[8:8+(f2&0xF)]
            ret.append((address, f2>>16, dddat, (f2>>4)&0xFF))
        return ret
